lib/BuildSystem/BuildSystem.cpp

```diff
@@ -380,9 +381,9 @@ class TargetTask : public Task {
   // FIXME: We should probably factor this out somewhere else, so we can enforce
   // it is never used when initialized incorrectly.
 
-  /// If true, the command had a missing input (this implies ShouldSkip is
-  /// true).
-  bool hasMissingInput = false;
+  /// If there are any elements, the command had missing inputs with the given
+  /// names (this implies ShouldSkip is true).
+  std::vector<StringRef> namesOfMissingInputs;
```

Comment:
Storing a StringRef is sketchy, are you sure its lifetime won't extend past the data that is provided? Typically storing a std::string is safer.

Also, should this be an ordered set data structure (we have one in LLVM) rather than a vector?

Reply:
It would be good to use one of LLVM's Small* data structures here if possible, to avoid an unnecessary allocation (not particularly important for `TargetTask`, but it matters some for `ExternalCommand`).

Reply:
True, I'll take your other suggestion and just store the node instead.